File: tests/openwrt.rs
//! OpenWrt adapter integration tests: each scenario spawns a fake ubus
//! endpoint (tests/fake_openwrt.py — real JSON-RPC shapes, real session-id
//! enforcement) on a free port plus the real supervisor on the openwrt
//! fixture house, and asserts on the bus. The fixture polls every second
//! with a 2s presence away-delay, so transitions land within the timeouts.

mod common;

use std::io::{Read, Write};
use std::path::PathBuf;
use std::process::{Child, Command, Stdio};
use std::time::{Duration, Instant};

use serde_json::{json, Value};
use zenoh::sample::SampleKind;

use common::{expect_event_kind, expect_state, fixture_command, free_port, Supervisor};

const FIXTURE: &str = "tests/fixture_house_openwrt";
const ROUTERS_ENV: &str = "HOMEOSTAT_OPENWRT";
const EVENT_KEY: &str = "home/health/openwrt/event";
const WAN_KEY: &str = "home/state/hallway/gateway/wan";
const PRESENCE_KEY: &str = "home/state/global/dads_phone/presence";
const DISCOVERY_KEY: &str = "home/discovery/openwrt";
const PHONE_MAC: &str = "aa:bb:cc:dd:ee:ff";
const USERNAME: &str = "homeostat";
const PASSWORD: &str = "secret123";

/// A fake ubus endpoint (tests/fake_openwrt.py) on a free port, killed on
/// drop. Spawned the same way the units themselves are: out of the
/// script's own uv environment, exec'd directly (fixture_command, #140),
/// so the handle held here is the interpreter's and not uv's.
struct FakeOpenwrt {
    child: Child,
    port: u16,
}

impl FakeOpenwrt {
    async fn spawn() -> Self {
        let port = free_port();
        let (program, args) = fixture_command(&format!(
            "uv run tests/fake_openwrt.py --port {port} --username {USERNAME} --password {PASSWORD}"
        ))
        .await;
        let child = Command::new(program)
            .args(args)
            .current_dir(env!("CARGO_MANIFEST_DIR"))
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
            .expect("spawn fake openwrt router (is uv installed?)");
        // First run resolves the fake router's own uv env: generous.
        let deadline = Instant::now() + Duration::from_secs(60);
        while std::net::TcpStream::connect(("127.0.0.1", port)).is_err() {
            assert!(
                Instant::now() < deadline,
                "fake openwrt router never listened on {port}"
            );
            std::thread::sleep(Duration::from_millis(50));
        }
        Self { child, port }
    }

    /// POSTs a control endpoint (plain HTTP, out of the JSON-RPC path).
    fn control(&self, path: &str) {
        let mut stream = std::net::TcpStream::connect(("127.0.0.1", self.port))
            .expect("connect to fake router control");
        stream
            .write_all(
                format!(
                    "POST {path} HTTP/1.1\r\nHost: 127.0.0.1\r\nContent-Length: 0\r\nConnection: close\r\n\r\n"
                )
                .as_bytes(),
            )
            .expect("write control request");
        let mut response = String::new();
        stream
            .read_to_string(&mut response)
            .expect("read control response");
        assert!(
            response.starts_with("HTTP/1.1 200"),
            "control {path}: {response}"
        );
    }
}

impl Drop for FakeOpenwrt {
    fn drop(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

/// Writes a HOMEOSTAT_OPENWRT file (outside the repo, per the settlement)
/// giving each named router a fake endpoint's host and the read-only rpcd
/// credentials. The fixture binds entities on "gw"; any further router is
/// configured but unbound, which is all a second AP needs to be polled.
fn routers_file(routers: &[(&str, u16)]) -> PathBuf {
    let tag: Vec<String> = routers.iter().map(|(_, port)| port.to_string()).collect();
    let path = std::env::temp_dir().join(format!("homeostat-openwrt-{}.toml", tag.join("-")));
    let body: String = routers
        .iter()
        .map(|(name, port)| {
            format!(
                "[{name}]\nhost = \"127.0.0.1:{port}\"\nusername = \"{USERNAME}\"\npassword = \"{PASSWORD}\"\n"
            )
        })
        .collect();
    std::fs::write(&path, body).expect("write routers file");
    path
}

/// Spawns the fake router + supervisor on the fixture and waits for the
/// adapter's liveliness token (generous timeout: first run resolves the
/// uv env for aiohttp too).
async fn setup() -> (FakeOpenwrt, PathBuf, Supervisor, zenoh::Session) {
    let router = FakeOpenwrt::spawn().await;
    let routers_path = routers_file(&[("gw", router.port)]);
    let (sup, observer) = start(&routers_path).await;
    (router, routers_path, sup, observer)
}

/// The supervisor on the fixture against a written routers file, up to the
/// adapter's liveliness token.
async fn start(routers_path: &std::path::Path) -> (Supervisor, zenoh::Session) {
    let sup = Supervisor::spawn_with_env(
        FIXTURE,
        &[(ROUTERS_ENV, routers_path.to_str().expect("utf-8 path"))],
    );
    let observer = sup.observer().await;
    let token_sub = observer
        .liveliness()
        .declare_subscriber("home/health/openwrt/alive")
        .history(true)
        .await
        .expect("liveliness subscriber");
    let token = tokio::time::timeout(Duration::from_secs(90), token_sub.recv_async())
        .await
        .expect("adapter liveliness token within 90s")
        .expect("liveliness stream open");
    assert_eq!(token.kind(), SampleKind::Put);
    (sup, observer)
}

/// (a) A phone associating to an AP becomes `presence = true` on the bus;
/// dropping off flips it back after the away delay — and the discovery
/// feed shows the sighted station bound to its entity.
#[tokio::test(flavor = "multi_thread")]
async fn wifi_association_drives_presence() {
    let (router, _routers_path, mut sup, observer) = setup().await;
    let presence_sub = observer
        .declare_subscriber(PRESENCE_KEY)
        .await
        .expect("presence subscriber");
    let discovery_sub = observer
        .declare_subscriber(DISCOVERY_KEY)
        .await
        .expect("discovery subscriber");

    router.control(&format!("/control/station?mac={PHONE_MAC}&present=true"));
    expect_state(&presence_sub, json!(true)).await;

    // The station's appearance changes the periphery: one JSON array,
    // the phone's record bound to its entity.
    let deadline = tokio::time::Instant::now() + Duration::from_secs(20);
    loop {
        let sample = tokio::time::timeout_at(deadline, discovery_sub.recv_async())
            .await
            .expect("discovery publish within 20s")
            .expect("discovery stream open");
        let records: Value = serde_json::from_slice(&sample.payload().to_bytes())
            .expect("discovery payload is JSON");
        let phone = records
            .as_array()
            .expect("discovery payload is an array")
            .iter()
            .find(|r| r["id"] == PHONE_MAC);
        if let Some(record) = phone {
            assert_eq!(record["configured"], json!(true), "phone record: {record}");
            assert_eq!(
                record["entity"],
                json!("dads_phone"),
                "phone record: {record}"
            );
            // and its aspect descriptor: the one boolean this capability speaks
            assert_eq!(
                record["aspects"]["fields"]["presence"]["kind"],
                json!("boolean"),
                "{record}"
            );
            assert_eq!(
                record["aspects"]["fields"]["presence"]["values"][1]["label"],
                json!("away")
            );
            break;
        }
    }

    router.control(&format!("/control/station?mac={PHONE_MAC}&present=false"));
    expect_state(&presence_sub, json!(false)).await;

    sup.shutdown();
}

/// (b) Connectivity state: WAN down is a `wan = false` transition, and
/// back up again. Tunnels left with the `vpn` capability (design.md,
/// amended 2026-09-23): this adapter reports presence and WAN.
#[tokio::test(flavor = "multi_thread")]
async fn connectivity_state_translates_to_bus() {
    let (router, _routers_path, mut sup, observer) = setup().await;
    let wan_sub = observer
        .declare_subscriber(WAN_KEY)
        .await
        .expect("wan subscriber");

    router.control("/control/wan?up=false");
    expect_state(&wan_sub, json!(false)).await;

    router.control("/control/wan?up=true");
    expect_state(&wan_sub, json!(true)).await;

    sup.shutdown();
}

/// (c) An unreachable router emits one "router-unreachable" health event
/// and its aspects go stale rather than false; polling resumes after
/// recovery without a restart.
#[tokio::test(flavor = "multi_thread")]
async fn unreachable_router_drops_once_and_recovers() {
    let (router, _routers_path, mut sup, observer) = setup().await;
    let wan_sub = observer
        .declare_subscriber(WAN_KEY)
        .await
        .expect("wan subscriber");
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");

    router.control("/control/break");
    expect_event_kind(&event_sub, "router-unreachable").await;

    router.control("/control/restore");
    router.control("/control/wan?up=false");
    expect_state(&wan_sub, json!(false)).await;

    sup.shutdown();
}

/// (e) A router whose replies stream is read whole. aiohttp's
/// `content.read(n)` hands back only what is buffered — the first chunk of
/// a chunked reply — so a single read truncates the JSON and every decode
/// fails. rpcd itself answers with Content-Length, which is why this has
/// never bitten in production here; the identical read in the onvif
/// adapter faced firmware that does stream, and took both of VP52's
/// cameras down on v0.12.0. The fix belongs to both call sites, so the
/// test does too.
///
/// Chunking is switched on mid-run, after presence has already been proved
/// to work, so the assertion is about the framing and nothing else.
#[tokio::test(flavor = "multi_thread")]
async fn a_chunked_ubus_reply_is_read_whole() {
    let (router, _routers_path, mut sup, observer) = setup().await;
    let presence_sub = observer
        .declare_subscriber(PRESENCE_KEY)
        .await
        .expect("presence subscriber");

    // The control: unchunked, presence follows the station.
    router.control(&format!("/control/station?mac={PHONE_MAC}&present=true"));
    expect_state(&presence_sub, json!(true)).await;

    router.control("/control/chunked");
    router.control(&format!("/control/station?mac={PHONE_MAC}&present=false"));
    expect_state(&presence_sub, json!(false)).await;

    sup.shutdown();
}

/// (f) Partial blindness is not absence. Two routers configured, the phone
/// associated only to the AP: when the AP goes silent the phone must hold
/// stale, because the union of the routers that *did* answer says nothing
/// about a device that lives on the one that did not — VP52 published
/// `presence = false` for 37 hours with the family at home (#144). The
/// blind spot is announced, and absence becomes assertable again once
/// every router answers.
#[tokio::test(flavor = "multi_thread")]
async fn a_silent_router_cannot_assert_absence() {
    let gw = FakeOpenwrt::spawn().await;
    let ap = FakeOpenwrt::spawn().await;
    let routers_path = routers_file(&[("gw", gw.port), ("ap", ap.port)]);
    let (mut sup, observer) = start(&routers_path).await;
    let presence_sub = observer
        .declare_subscriber(PRESENCE_KEY)
        .await
        .expect("presence subscriber");
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");

    ap.control(&format!("/control/station?mac={PHONE_MAC}&present=true"));
    expect_state(&presence_sub, json!(true)).await;

    ap.control("/control/break");
    expect_event_kind(&event_sub, "presence-partial").await;

    // Several poll cycles past the 2s away delay: nothing may say away.
    let deadline = tokio::time::Instant::now() + Duration::from_secs(8);
    while let Ok(sample) = tokio::time::timeout_at(deadline, presence_sub.recv_async()).await {
        let value: Value =
            serde_json::from_slice(&sample.expect("state stream open").payload().to_bytes())
                .expect("state payload is JSON");
        assert_ne!(
            value,
            json!(false),
            "presence asserted away while the AP that sees the phone was silent"
        );
    }

    // The AP answering again, the phone really gone: away, as before.
    ap.control("/control/restore");
    ap.control(&format!("/control/station?mac={PHONE_MAC}&present=false"));
    expect_state(&presence_sub, json!(false)).await;

    sup.shutdown();
}

File: adapters/openwrt.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "homeostat",
#     "aiohttp>=3.12.14,<4",
# ]
#
# [tool.uv.sources]
# homeostat = { path = "../sdk/python", editable = true }
# ///
"""OpenWrt network adapter (see docs/design.md, "Network presence and
connectivity (settled 2026-07-25)").

Named for the dialect it speaks: ubus JSON-RPC over HTTP (uhttpd-mod-ubus,
rpcd session auth). The first polling adapter — each cycle logs in fresh
(rpcd expires idle sessions; nothing to renew) and asks every configured
router two questions: `network.interface dump` for WAN state and
`get_clients` on every hostapd BSS for WiFi sightings. Scope is presence
and WAN state only; network metrics are the monitoring stack's job,
deliberately, and tunnel state left with the `vpn` capability
(docs/design.md, amended 2026-09-23).

The manifest's [discovery].endpoint is the HOMEOSTAT_OPENWRT credentials
file itself: an out-of-repo TOML keyed by router name with `host`
(optionally `host:port`), `username`, `password` — a dedicated read-only
rpcd ACL login per router, never root. Entity binding: a `router` entity's
id is its name in that file (aspect `wan`); a `presence` entity's id is the device MAC,
lowercase (aspect `presence` — sighted on any BSS of any router, absent
only after away_delay_s of continuous non-sighting, which also absorbs AP
reboots, and only while every configured router polled: a silent router is
a blind spot, not an empty one).

Aspects publish on transition only, plus each entity's current value after
its first successful poll; a poll is a read, not an event. An unreachable
router drops with one "router-unreachable" health event per down
transition and its aspects go stale rather than false. Read-only by
design: no cmd surface until a command is actually wanted.
A bound entity's discovery record carries its aspect descriptor (docs/
design.md, Aspect descriptors): one boolean per capability, with value
labels ("up"/"down", "present"/"away") — the whole of what this adapter
speaks, so a static table, not a generated one.
"""

import asyncio
import contextlib
import json
import os
import signal
import time
from pathlib import Path
from typing import ClassVar

import aiohttp
import homeostat
import tomllib
from homeostat import house, keys
from homeostat.params import LiveParams

NULL_SID = "0" * 32
HTTP_TIMEOUT_S = 10
PARAM_DEFAULTS = {"poll_interval_s": 30, "away_delay_s": 180}


class UbusError(Exception):
    """Any failure of a ubus round trip: HTTP status, JSON-RPC error,
    non-zero ubus status code, unparseable body."""


# A ubus reply is a small JSON document; a compromised or misbehaving
# router must not pin the unit's memory on an oversized one.
MAX_RESPONSE_BYTES = 1024 * 1024
# How much of the body one read takes. Only a buffer size: the cap above
# is what bounds memory, and it is checked after every chunk.
RESPONSE_CHUNK_BYTES = 64 * 1024


async def ubus_rpc(http: aiohttp.ClientSession, url: str, method: str, params: list):
    payload = {"jsonrpc": "2.0", "id": 1, "method": method, "params": params}
    try:
        async with http.post(
            url, json=payload, timeout=aiohttp.ClientTimeout(total=HTTP_TIMEOUT_S),
            # A redirect would replay the login call (the plaintext rpcd
            # password included) at whatever host the reply names.
            allow_redirects=False,
        ) as response:
            if response.status != 200:
                raise UbusError(f"HTTP {response.status}")
            # ⚠️ READ UNTIL EOF, NOT ONCE. `content.read(n)` returns
            # whatever is buffered, up to n -- for a chunked reply that is
            # the FIRST CHUNK, so a single read truncates the document and
            # every decode fails. rpcd here does answer chunked (OpenWrt
            # 23.x, bodies of 1.4-5.7 kB), and a single read took VP52's
            # AP off the air for two days (#144) -- this was never the
            # insurance it was first described as. The cap is still enforced,
            # now after each chunk, which is also where it belongs -- it
            # must not depend on how the body happens to be framed.
            raw = bytearray()
            async for chunk in response.content.iter_chunked(RESPONSE_CHUNK_BYTES):
                raw += chunk
                if len(raw) > MAX_RESPONSE_BYTES:
                    raise UbusError(f"response exceeds {MAX_RESPONSE_BYTES} bytes")
            data = json.loads(bytes(raw))
    except aiohttp.ClientError as err:
        raise UbusError(str(err)) from err
    except asyncio.TimeoutError as err:
        raise UbusError("timeout") from err
    except ValueError as err:
        raise UbusError(f"unparseable response: {err}") from err
    if not isinstance(data, dict) or "error" in data:
        raise UbusError(f"rpc error: {data.get('error') if isinstance(data, dict) else data}")
    return data.get("result")


async def ubus_call(http, url: str, sid: str, obj: str, method: str, args: dict) -> dict:
    result = await ubus_rpc(http, url, "call", [sid, obj, method, args])
    if not isinstance(result, list) or not result or result[0] != 0:
        raise UbusError(f"{obj}.{method}: status {result!r}")
    return result[1] if len(result) > 1 else {}


async def ubus_list(http, url: str, sid: str, pattern: str) -> list[str]:
    """Object names matching a pattern (hostapd.* — one per BSS)."""
    result = await ubus_rpc(http, url, "list", [sid, pattern])
    if isinstance(result, dict):
        return sorted(result)
    if isinstance(result, list) and result and isinstance(result[-1], dict):
        return sorted(result[-1])
    return []


async def login(http, url: str, username: str, password: str) -> str:
    data = await ubus_call(
        http, url, NULL_SID, "session", "login", {"username": username, "password": password}
    )
    sid = data.get("ubus_rpc_session")
    if not sid:
        raise UbusError("login reply missing ubus_rpc_session")
    return sid


class Params(LiveParams):
    """poll_interval_s / away_delay_s from home/config/{unit}/*, live."""

    @property
    def poll_interval_s(self) -> float:
        return max(1, self.get("poll_interval_s"))

    @property
    def away_delay_s(self) -> float:
        return max(0, self.get("away_delay_s"))


def load_routers(endpoint: str | None) -> dict:
    """The HOMEOSTAT_OPENWRT TOML behind [discovery].endpoint: per-router
    host/username/password keyed by router name. Missing or unreadable is
    a startup error (visible via the supervisor's backoff)."""
    if not endpoint:
        raise ValueError("openwrt adapter requires [discovery].endpoint")
    return tomllib.loads(Path(endpoint).read_text())


def classify(entities, routers, session):
    """Splits bound entities by capability, dropping unusable bindings
    with a health event each — one bad entity never takes the unit down."""
    router_entities, trackers = [], []
    for entity in entities:
        if entity.capability == "router":
            if entity.id not in routers:
                session.health_event("drop", reason="router-unconfigured", entity=entity.name)
                continue
            router_entities.append(entity)
        elif entity.capability == "presence":
            if entity.id != entity.id.lower():
                # Sightings are lowercased; an uppercase MAC would just
                # read absent forever with no trace.
                session.health_event(
                    "drop", reason="malformed-id", entity=entity.name,
                    error="device MAC must be lowercase",
                )
                continue
            trackers.append(entity)
        else:
            session.health_event(
                "drop", reason="unsupported-capability", entity=entity.name,
                capability=entity.capability,
            )
    return router_entities, trackers


async def poll_router(http, name: str, conf: dict):
    """One router, one cycle: fresh login, interface dump, station union.
    Any failure marks the router unreachable."""
    url = f"http://{conf['host']}/ubus"
    sid = await login(http, url, conf["username"], conf["password"])

    dump = await ubus_call(http, url, sid, "network.interface", "dump", {})
    interfaces = {
        iface.get("interface"): iface
        for iface in dump.get("interface", [])
        if iface.get("interface")
    }

    stations: set[str] = set()
    for obj in await ubus_list(http, url, sid, "hostapd.*"):
        clients = await ubus_call(http, url, sid, obj, "get_clients", {})
        stations |= {mac.lower() for mac in clients.get("clients", {})}

    return interfaces, stations


class Adapter:
    def __init__(self, session, routers, router_entities, trackers):
        self.session = session
        self.routers = routers
        self.router_entities = router_entities
        self.trackers = trackers
        self.params = Params(session, PARAM_DEFAULTS)
        self.published: dict[str, object] = {}
        self.last_discovery: str | None = None
        self.reachable: dict[str, bool] = {}
        self.noted: set = set()  # degraded conditions already announced
        self.blind: frozenset = frozenset()  # routers silent as of last cycle
        self.last_seen: dict[str, float] = {}  # mac -> monotonic sighting time

    def note(self, key: tuple, kind: str, **fields) -> None:
        """One health event per down transition of a degraded condition.
        Degraded conditions publish kind = condition (the recorder's
        backend-outage precedent) — nothing was dropped."""
        if key not in self.noted:
            self.session.health_event(kind, **fields)
            self.noted.add(key)

    def clear(self, key: tuple) -> None:
        self.noted.discard(key)

    def publish(self, key: str, value) -> None:
        if self.published.get(key) != value:
            self.session.put_json(key, value)
            self.published[key] = value

    async def cycle(self, http) -> None:
        interfaces: dict[str, dict] = {}
        sightings: dict[str, list[str]] = {}  # mac -> routers that saw it
        polled_ok: set[str] = set()

        for name, conf in self.routers.items():
            try:
                ifaces, stations = await poll_router(http, name, conf)
            except UbusError as err:
                if self.reachable.get(name, True):
                    self.session.health_event(
                        "router-unreachable", router=name, error=str(err)
                    )
                self.reachable[name] = False
                continue
            except Exception as err:
                # A payload shape this build did not anticipate (hostapd
                # clients as a list, an interface dump that is not one, ...): the
                # router answered, the answer just did not parse. Stale, not
                # a crash loop — one bad router never takes the unit down.
                if self.reachable.get(name, True):
                    self.session.health_event(
                        "router-poll-failed", router=name, error=str(err)
                    )
                self.reachable[name] = False
                continue
            self.reachable[name] = True
            polled_ok.add(name)
            interfaces[name] = ifaces
            for mac in stations:
                sightings.setdefault(mac, []).append(name)

        if not polled_ok:
            return  # total blindness: everything stays stale

        now_mono = time.monotonic()
        for mac in sightings:
            self.last_seen[mac] = now_mono

        for entity in self.router_entities:
            if entity.id not in polled_ok:
                continue
            wan = interfaces[entity.id].get("wan")
            if wan is None:
                self.note(("iface", entity.name), "unknown-interface", entity=entity.name, iface="wan")
                continue
            self.clear(("iface", entity.name))
            self.publish(keys.state_key(entity.room, entity.name, "wan"), bool(wan.get("up")))

        # Partial blindness. A sighting is evidence whoever else failed,
        # but absence is the union of all routers seeing nothing -- with
        # one silent, a device that lives on it would read away. Hold
        # those stale, exactly as the total-blindness branch does, and
        # announce the blind spot per change of which routers are silent
        # (router-unreachable is latched once and says nothing about how
        # long the outage runs).
        blind = frozenset(self.routers) - polled_ok
        if self.trackers and blind != self.blind and blind:
            self.session.health_event("presence-partial", routers=sorted(blind))
        self.blind = blind

        away_delay = self.params.away_delay_s
        for entity in self.trackers:
            seen_at = self.last_seen.get(entity.id)
            present = entity.id in sightings or (
                seen_at is not None and now_mono - seen_at < away_delay
            )
            if not present and blind:
                continue
            self.publish(keys.state_key(entity.room, entity.name, "presence"), present)

        self.publish_discovery(sightings)

    # A class-level constant, shared read-only across instances — never
    # mutated per router.
    ASPECT_DESCRIPTORS: ClassVar[dict] = {
        "router": {
            "schema": 1,
            "groups": ["readings"],
            "fields": {
                "wan": {
                    "label": "WAN link (wan)", "kind": "boolean", "group": "readings",
                    "values": [{"value": True, "label": "up"}, {"value": False, "label": "down"}],
                }
            },
        },
        "presence": {
            "schema": 1,
            "groups": ["readings"],
            "fields": {
                "presence": {
                    "label": "on the WiFi (presence)", "kind": "boolean", "group": "readings",
                    "values": [{"value": True, "label": "present"}, {"value": False, "label": "away"}],
                }
            },
        },
    }

    def publish_discovery(self, sightings) -> None:
        """The complete current view of the periphery (docs/design.md,
        Discovery), from data the cycle already fetched; republished only
        when it changes."""
        bound = {e.id: e.name for e in self.router_entities + self.trackers}

        def record(rid: str, capability: str, description: dict) -> dict:
            rec = {
                "id": rid,
                "configured": rid in bound,
                "entity": bound.get(rid),
                "suggested": {"capability": capability, "features": []},
                "description": description,
            }
            if rid in bound:
                rec["aspects"] = self.ASPECT_DESCRIPTORS[capability]
            return rec

        records = [
            record(name, "router", {"reachable": self.reachable.get(name, False)})
            for name in self.routers
        ]
        for mac in sorted(sightings):
            records.append(record(mac, "presence", {"seen_by": sorted(sightings[mac])}))

        records.sort(key=lambda r: r["id"])
        serialized = json.dumps(records)
        if serialized != self.last_discovery:
            self.session.put_json(keys.discovery_key(self.session.unit), records)
            self.last_discovery = serialized


async def serve(session, routers, config) -> None:
    loop = asyncio.get_running_loop()
    stop = asyncio.Event()
    for sig in (signal.SIGTERM, signal.SIGINT):
        loop.add_signal_handler(sig, stop.set)

    adapter = Adapter(session, routers, *classify(config.entities, routers, session))

    async with aiohttp.ClientSession() as http:
        first = True
        while not stop.is_set():
            await adapter.cycle(http)
            if first:
                # The first cycle has run (reachable or not — its own loop
                # keeps trying); the unit is wired up.
                session.ready()
                first = False
            with contextlib.suppress(asyncio.TimeoutError):
                await asyncio.wait_for(stop.wait(), timeout=adapter.params.poll_interval_s)


def main() -> None:
    unit = os.environ[keys.ENV_UNIT]
    config = house.load_adapter(unit)
    routers = load_routers(config.endpoint)

    session = homeostat.connect()
    try:
        asyncio.run(serve(session, routers, config))
    finally:
        session.close()


if __name__ == "__main__":
    main()
